Feature/logging

Joshua-Lawson opened the pull request:
Changes to improve logging within the Autonomy system, and increase communication with other systems
- Implements the python logging library to improve logging
- Uses a YAML config file to adjust logging settings
- Adapted Rovecomm TCP to send packets to all connected sockets, similar to UDP subscriptions
- Adds a logging handler to send log strings over Rovecomm to subscribers
- Adds a telemetry handler function to send numerical data over Rovecomm to subscribers
- Added a JSON dictionary to template numerical telemetry that will be sent out
- Includes a small script to demonstrate reading test data from logs

Commit:
Playing around with logging.py

Includes handlers to send data over RoveComm through both UDP and TCP
Includes a handler to parse specific logs for numerical data to send over RoveComm

Commit:
Added CSV handler

Also fixed a couple file reference errors caused by moving folders

Commit:
Merge branch 'feature/runner' into feature/logging

Commit:
Moved csv logs to their own folder

Commit:
Moved logging to a unique name, changed run.py to use yaml config

Commit:
Changed logging's main.py to logging_test.py and added it to run.py

Commit:
Merge branch 'dev' into feature/logging

Commit:
Moved filters to their own file, set logging yaml to match current logging setup

Commit:
Formatting, changed csv handler to name based on date

Commit:
Condensed and moved contents of logging_files to core

Commit:
Moved numerical rovecomm data to its own function with parameters

Commit:
Removed functions that supported numerical data as a log handler

Commit:
Fixed problems caused by renaming the file

Commit:
Merge branch 'dev' into feature/logging

Commit:
Adapted to new rovecomm implementation

Commit:
improved comments

Commit:
Added string datatype to rovecomm

Commit:
added unit test for rovecomm logging

Commit:
fixed logger error in rovecomm

Commit:
Fixed closing log check

Commit:
Closes rovecomm thread at end of program

Commit:
Added float datatype to rovecomm to match cpp

Commit:
Allowed csv handler to work with a single output file

Commit:
Merge branch 'dev' into feature/logging

Commit:
Fixed unit test to match dev changes

Commit:
Wraps message in quotes to escape any contained commas

Commit:
Improved csv format, added an example log reading file

Commit:
commenting

Commit:
Converted event list to json

Commit:
Set rovecomm handlers to INFO and above

Commit:
Added more state values to json file

Commit:
Set rovecomm tcp to send to all open sockets

Commit:
Set packets port to 0 to only send to subscribers

Commit:
Adapted the logging unit test to sending to subscribers

Commit:
Added pyyaml dependency to pipfile

Commit:
Renamed basestation_send to telemetry_handler

Commit:
Removed the unused file handler

Commit:
Populated rovecomm_values with more telemetry

Commit:
Improved docstring for read_logs main

Commit:
cleaned up read_logs.py

Commit:
Removed CLoader

CLoader was recommended as an optimization for speed, but is causing issues in the build process

Commit:
csv handler now creates necessary folders to contain logs

Commit:
Changed TCP to allow the port to be reused immediately

thomas-gleiforst commented on core/rovecomm.py:
```diff
@@ -22,8 +22,10 @@
+    6: 'f',
+    7: 'q',  # int64, this needs to stay here to not break RED. Leave this comment here Skelton.
+    8: 'd',  # double
+    9: 's',
```
Think about changing this to "c" to make it a little more transparent this is a character array

thomas-gleiforst commented on core/logging_handlers.py:
```diff
@@ -0,0 +1,68 @@
+            msg = msg[:252] + "..."
+
+        packet = RoveCommPacket(
+            4241,
```
make dataid a config in the yaml file rather than hardcoded in this obscure place

thomas-gleiforst commented on core/logging_handlers.py:
```diff
@@ -0,0 +1,68 @@
+
+class CsvHandler(logging.handlers.WatchedFileHandler):
+
+    def __init__(self, filename, format_string, encoding=None, delay=False, new_file=False):
```
change parameter names to make functionality more clear

thomas-gleiforst commented on core/rovecomm.py:
```diff
@@ -381,13 +391,17 @@ def write(self, packet):
-            if self.connect(packet.ip_address) == 0:
-                return 0
+            for address in self.open_sockets:
+                self.open_sockets[address].send(rovecomm_packet)
```
We should split off open connections we initiate and open connections that were requested from us, and then only send packets to people who requested from us. That was we don't do something like send logs to drive board.

thomas-gleiforst commented on read_logs.py:
```diff
@@ -0,0 +1,36 @@
+import logging
+import csv
+from datetime import datetime
+import matplotlib.pyplot as plt
+
+
+def main() -> None:
+    '''
+    Reads some telemetry from the logs as an example of post-test analysis
+    '''
+    logger = logging.getLogger(__name__)
+    logger.info("Executing function: main()")
+
+    log_file = "logs/newformat-20201024-133704.csv"
+
+    # Gets motor differential data
+    timestamps = []
+    ratios = []
+    with open(log_file) as csvfile:
+        reader = csv.DictReader(csvfile)
+        for row in reader:
+            if row['MESSAGE'].startswith('Driving at '):
+                speeds = eval(row['MESSAGE'][11:])
+                ratios.append(speeds[0] - speeds[1])
+                timestamps.append(datetime.strptime(row['TIMESTAMP'], "%Y-%m-%d %H:%M:%S.%f"))
+
+    # Graph the data with matplot
+    fig, ax = plt.subplots()
+    ax.plot(timestamps, ratios)
+    plt.title("Motor Differential over Time")
+    plt.show()
+
+
+if __name__ == "__main__":
+    # Run main()
+    main()
```
Consider moving this to an example script folder so we can keep this as a reference in case we want to do some similar analysis of logs later, but shouldn't clutter the main directory since it will almost never be run in its current form.

thomas-gleiforst commented on run.py:
```diff
@@ -16,37 +18,11 @@ def setup_logger(level) -> logging.Logger:
-    logger = logging.getLogger()
-    logger.setLevel(logging.DEBUG)
-    
-    # Primary handlers. Levels can be set with --level option
```
Consider readding a level, especially for console logging, for cases where we only want to see critical and error logging, not absolutely everything.

thomas-gleiforst commented on core/logging_filters.py:
```diff
@@ -0,0 +1,6 @@
+import logging
+
+
+class DebugFilter(logging.Filter):
+    def filter(self, record):
+        return (record.levelno == logging.DEBUG)
```
This filter doesn't appear to be used anywhere, and you should consider removing or re-justifying it.

thomas-gleiforst commented:
Missed a comment: you also should look into changing the rovecomm testing script because you made changes to the way TCP sends to multiple subscribers and didn't update the test.

Commit:
Cleaned up some variable names

Commit:
Wrapping up pull request changes

Commit:
Tries the REUSEPORT setting if it's available

Commit:
Added incoming_sockets as a subset of open_sockets

Now connections that are made from this end won't be considered subscribed to all packets that are sent
